### PR TITLE
Persist online sessions to the database

### LINKED ISSUE #27976: Persist online sessions to the database
### Description

While Keycloak's offline sessions are stored in the database, the online sessions are only stored in Infinispan, and are not persisted during restarts and upgrades.

### Motivation

Enable more seamless upgrades between Keycloak releases. Eventually allow having not all user sessions in memory.

Keycloak already has the tables `offline_user_session` / `offline_client_session` to store this information. 

The `InfinispanChanglogBasedTransaction` already contains the mechanisms to replay changes to those sessions against other stores, which I'll leverage to run the changes against JPA.

### Details

I will create a PoC for this

### PR DESCRIPTION
**Pull request by ahus1**
Closes #27976

Additional overhead created by this PR compared to Infinispan: 
* When searching for sessions by users/clients/brokersessionid, it queries the database and not Infinispan. 
* All search results are imported to the embedded caches.
* When sessions are created or changed, the information is written to the database in a separate transactions (one for usersessions, one for client sessions). The positive side is that it allows for retrying the write and handle concurrency similar to the Infinispan handling. The negative side is that it is slower than writing it in the original DB transactions. 
* Additional option to completely disable the caches for online user and client sessions, and instead loading the data on all requests. This leverages the database cache. In X-Site environment, it avoid the expensive x-site request to the other site.

Example command line: 

```
bin/kc.sh start --http-enabled true --hostname-strict false --db postgres --db-password pass --db-username keycloak --features=persistent-user-sessions,persistent-user-sessions-no-cache
```

Optimizations (pending if this PR or another): 
* After writing to the database, always put that version into Infinispan to avoid inconsistencies (when writing, ensure no-one else has up an already more up-to-date version in there) - use Hibernate version numbering to determine the latest entry.

Future optimizations (outside of the scope of this PR): 
* Import existing online sessions (which might be stored in a remote or embedded Infinispan) to the database - this might enable users of large setups to evaluate this)
* Write user and client session in the same DB transaction
* On refreshing of tokens, update the entries with their refresh time on a per-second basis, but possibly more granular (every 15 min?) to reduce database updates

### CODE REVIEW

**Comment by kami619** [comment]
**Performance test with a ROSA cross-site setup:**

KC_INSTANCES=3
KC_CPU_REQUESTS: 6
KC_INSTANCES: 3
KC_DISABLE_STICKY_SESSION: true
KC_MEMORY_REQUESTS_MB: 3000
KC_MEMORY_LIMITS_MB: 4000
KC_DB_POOL_INITIAL_SIZE: 30
KC_DB_POOL_MAX_SIZE: 30
KC_DB_POOL_MIN_SIZE: 30
KC_DATABASE: aurora-postgres

**Gatling test run (using 0.13-SNAPSHOT)**
```
./benchmark.sh eu-west-1 --scenario=keycloak.scenario.authentication.AuthorizationCode \
--server-url=https://client.gh-keycloak-a-gh-keycloak-b-otm0cg.keycloak-benchmark.com \
--realm-name=realm-0 --users-per-sec=40 --ramp-up=20 --logout-percentage=0 \
--measurement=600 --users-per-realm=20000 --log-http-on-failure --sla-error-percentage=0.001

```

**Results:**
Screenshot 2024-03-20 at 15 44 32

[authorizationcode-20240320152628.zip](https://github.com/keycloak/keycloak/files/14671832/authorizationcode-20240320152628.zip)


**Observed Errors/Warns:**

OptimisticLockException: 
```
{"timestamp":"2024-03-20T19:33:45.011979968Z","sequence":26214,"loggerClassName":"com.arjuna.ats.arjuna.logging.arjunaI18NLogger_$logger","loggerName":"com.arjuna.ats.arjuna","level":"WARN","message":"ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff0a830406:b465:65fb3197:322b4, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@3981aa50 >","threadName":"executor-thread-42","threadId":237,"mdc":{},"ndc":"","hostName":"keycloak-2","processName":"QuarkusEntryPoint","processId":1,"exception":{"refId":1,"exceptionType":"jakarta.persistence.OptimisticLockException","message":"Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.keycloak.models.jpa.session.PersistentUserSessionEntity#org.keycloak.models.jpa.session.PersistentUserSessionEntity$Key@4c51bfe1]","frames":[{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"wrapStaleStateException","line":209},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":95},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":167},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":173},{"class":"org.hibernate.internal.SessionImpl","method":"doFlush","line":1407},{"class":"org.hibernate.internal.SessionImpl","method":"managedFlush","line":484},{"class":"org.hibernate.internal.SessionImpl","method":"flushBeforeTransactionCompletion","line":2319},{"class":"org.hibernate.internal.SessionImpl","method":"beforeTransactionCompletion","line":1976},{"class":"org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl","method":"beforeTransactionCompletion","line":439},{"class":"org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl","method":"beforeCompletion","line":336},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl","method":"beforeCompletion","line":47},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization","method":"beforeCompletion","line":37},{"class":"com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple","method":"beforeCompletion","line":52},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"beforeCompletion","line":351},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"end","line":69},{"class":"com.arjuna.ats.arjuna.AtomicAction","method":"commit","line":138},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple","method":"commitAndDisassociate","line":1271},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction","method":"commit","line":104},{"class":"io.quarkus.narayana.jta.runtime.NotifyingTransactionManager","method":"commit","line":70},{"class":"org.keycloak.transaction.JtaTransactionWrapper","method":"commit","line":90},{"class":"org.keycloak.services.DefaultKeycloakTransactionManager","method":"commit","line":136},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"closeTransactionManager","line":407},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"close","line":372},{"class":"org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction","method":"persistUserAndClientSession","line":541},{"class":"org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction","method":"lambda$commitImpl$1","line":202},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":102},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":89},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":80},{"class":"org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction","method":"commitImpl","line":202},{"class":"org.keycloak.models.AbstractKeycloakTransaction","method":"commit","line":48},{"class":"org.keycloak.services.DefaultKeycloakTransactionManager","method":"commit","line":146},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"closeTransactionManager","line":407},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"close","line":372},{"class":"org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler","method":"close","line":55},{"class":"org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionHandler","method":"closeSession","line":69},{"class":"org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionHandler","method":"filter","line":65},{"class":"org.jboss.resteasy.reactive.server.handlers.ResourceResponseFilterHandler","method":"handle","line":25},{"class":"io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext","method":"invokeHandler","line":150},{"class":"org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext","method":"run","line":147},{"class":"io.quarkus.vertx.core.runtime.VertxCoreRecorder$14","method":"runWith","line":582},{"class":"org.jboss.threads.EnhancedQueueExecutor$Task","method":"run","line":2513},{"class":"org.jboss.threads.EnhancedQueueExecutor$ThreadBody","method":"run","line":1522},{"class":"org.jboss.threads.DelegatingRunnable","method":"run","line":29},{"class":"org.jboss.threads.ThreadLocalResettingRunnable","method":"run","line":29},{"class":"io.netty.util.concurrent.FastThreadLocalRunnable","method":"run","line":30},{"class":"java.lang.Thread","method":"run","line":840}],"causedBy":{"exception":{"refId":2,"exceptionType":"org.hibernate.StaleObjectStateException","message":"Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.keycloak.models.jpa.session.PersistentUserSessionEntity#org.keycloak.models.jpa.session.PersistentUserSessionEntity$Key@4c51bfe1]","frames":[{"class":"org.hibernate.engine.jdbc.mutation.internal.ModelMutationHelper","method":"identifiedResultsCheck","line":75},{"class":"org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard","method":"lambda$doStaticUpdate$8","line":782},{"class":"org.hibernate.engine.jdbc.mutation.internal.ModelMutationHelper","method":"checkResults","line":50},{"class":"org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor","method":"performNonBatchedMutation","line":115},{"class":"org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched","method":"performNonBatchedOperations","line":40},{"class":"org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor","method":"execute","line":52},{"class":"org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard","method":"doStaticUpdate","line":778},{"class":"org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard","method":"performUpdate","line":324},{"class":"org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard","method":"coordinateUpdate","line":242},{"class":"org.hibernate.persister.entity.AbstractEntityPersister","method":"update","line":2810},{"class":"org.hibernate.action.internal.EntityUpdateAction","method":"execute","line":168},{"class":"org.hibernate.engine.spi.ActionQueue","method":"executeActions","line":632},{"class":"org.hibernate.engine.spi.ActionQueue","method":"executeActions","line":499},{"class":"org.hibernate.event.internal.AbstractFlushingEventListener","method":"performExecutions","line":363},{"class":"org.hibernate.event.internal.DefaultFlushEventListener","method":"onFlush","line":41},{"class":"org.hibernate.event.service.internal.EventListenerGroupImpl","method":"fireEventOnEachListener","line":127},{"class":"org.hibernate.internal.SessionImpl","method":"doFlush","line":1403},{"class":"org.hibernate.internal.SessionImpl","method":"managedFlush","line":484},{"class":"org.hibernate.internal.SessionImpl","method":"flushBeforeTransactionCompletion","line":2319},{"class":"org.hibernate.internal.SessionImpl","method":"beforeTransactionCompletion","line":1976},{"class":"org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl","method":"beforeTransactionCompletion","line":439},{"class":"org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl","method":"beforeCompletion","line":336},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl","method":"beforeCompletion","line":47},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization","method":"beforeCompletion","line":37},{"class":"com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple","method":"beforeCompletion","line":52},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"...

```

**Comment by ahus1** [comment]
@kami619 - thank you for this run. Still quite good for a first run. I see that on one of the nodes the CPU is at 100%. It seems that unfortunately two Keycloak nodes ended up on one node, and together they tried to eat more CPU than the node had with the given load. At the same time the DB connections got saturated, also quite expected, so as we discussed in the chat, let's reduce the load for now.

For a future run, if you would change the CPU requested to 9, this would ensure that each node only gets one Keycloak instance. Maybe we can look for more advanced measures to use some anti-affinity rules in the future, so they distribute evenly across nodes.

Still, this is beyond this set of runs. I think with all the data you collected I should try to do some optimizations first.

image

**Comment by kami619** [comment]
no probs

I have collected few more data points with 20 users-per-sec, 30 users-per-sec, 35 users-per-sec and 40 users-per-sec with Performance insights turned on. I will attempt a run where the cpu-affinity is better, sometime tomorrow and see if the connection handling is better. thank you.

based on these runs, I would say the OptimisticLocking exception is directly proportional to the run-rate.

[authCode_20ups.zip](https://github.com/keycloak/keycloak/files/14672878/authCode_20ups.zip)
[authCode_30ups.zip](https://github.com/keycloak/keycloak/files/14672879/authCode_30ups.zip)
[authCode_35ups.zip](https://github.com/keycloak/keycloak/files/14672880/authCode_35ups.zip)
[authCode_40ups_round2.zip](https://github.com/keycloak/keycloak/files/14672881/authCode_40ups_round2.zip)

**Comment by ahus1** [comment]
Thank you for these reports. I compared the 35 logins-per-second with the nightly run which is 40 logins-per-second.

For the "post correct credentials" step, there is a ~150ms extra latency for the request mean when we write to the multi-AZ Aurora database. All writes to the database are synchronously transferred to the secondary site, so they are expected to be slow. 

I am positively surprised that exchange code is of the same speed as before, I expected this to be slower as well.

In a future PR we could refactor the change to combine the changes to client and user session in a single transaction, which could reduce the overhead to 50%. 

image

image

**Comment by mhajas** [comment]
> I am positively surprised that exchange code is of the same speed as before, I expected this to be slower as well.

I would say this is expected as code-to-token is only reading sessions and not doing any writing to the database.

**Comment by ahus1** [comment]
> I am positively surprised that exchange code is of the same speed as before, I expected this to be slower as well.

I did a local test and found the following: if browser post and code exchange happen in the same second, the exchange code doesn't update the database. If it happens in two different seconds, it will update the database due to the change last refresh timestamp in the session (Hibernate skips the update as no data has been changed). With our gatling test it seems to happen in the same second almost every time. 

Again, another hint that we should reduce the granularity of the refresh time to reduce the number of database updates.

**Comment by pedroigor** [comment]
> I would say this is expected as code-to-token is only reading sessions and not doing any writing to the database.

But every time a refresh token is issued, both client and user sessions are updated.

**Comment by pedroigor** [comment]
> Again, another hint that we should reduce the granularity of the refresh time to reduce the number of database updates.

There is a trade-off here, right? The more you delay the refresh time, the higher the chance to expire sessions prematully, right?

Wouldn't such a setting introduce one more variable to take into account when deploying the server and potentially not helping much to actually make it (or keep the same) easier to deploy? I mean, what would be the default for this setting?

**Comment by ahus1** [comment]
@pedroigor 

> But every time a refresh token is issued, both client and user sessions are updated.

At the moment, we're updating the last-update-date with the accuracy of one second. 

> The more you delay the refresh time, the higher the chance to expire sessions prematully, right?
Wouldn't such a setting introduce one more variable to take into account when deploying the server and potentially not helping much to actually make it (or keep the same) easier to deploy? I mean, what would be the default for this setting?

Let's assume we would update the last-update-time with a granularity of 15 minutes. If access tokens have a max lifetime of 90 seconds and are refreshed by the clients after 60 seconds, this would reduce the load to 1/15th. 

We would probably implement it so we update the last-update-date with a time 15 minutes in the future, so it would rather extend the session idle time (but never the max lifetime), and ending a session prematurely.  

We would have a default (maybe 1, 5 or 15 minutes), so no-one would be forced to configure it. In the first iteration it might bei a SPI provider option you can tune. In the future there might a configuration in the UI per realm or per client.  

If we think this is not sufficient, we can also batch the updates, although I don't like the approach as I have seen the batches failing with deadlocks in the past, thereby loosing refreshes from all included items. And then the batches being triggered by ever 100th or so request when the queue was full, and this one request "paying" the delay for such a big batch which could block and even fail due to deadlocks. 

Tradoffs, tradeoffs everywhere! 

image

**Review by stianst — DISMISSED** [review]
I'd suggest we add this to a feature branch instead of merging to main for now.

**Comment by pedroigor** [comment]
> Tradoffs, tradeoffs everywhere!

Yeah, they are.

> We would have a default (maybe 1, 5 or 15 minutes), so no-one would be forced to configure it. In the first iteration it might bei a SPI provider option you can tune. In the future, there might a configuration in the UI per realm or per client.

If no-one will be forced to configure it, why do you want a setting for this then? 

Setting "1, 5, 15 minutes" implies a lot of trade-offs and, depending on your deployment, you might not want any timeout at all (like it works today).

In this first iteration, I would just focus on delivering the very basic needed to enable persistent sessions. Benchmark and performance could be done as a separate step. And then we can start discussing any trade-off you are introducing.

**Comment by pedroigor** [comment]
@ahus1 Btw, I'm not against having a timeout if that is the best we can do to improve throughput when persisting sessions. I'm just trying to make a point here that this will change how people deploy Keycloak and with consequences on how sessions are validated.

IMO, both of these points should be very clear for everybody in order to understand how people will actually use it. In an ideal world (which we don't live in), we would switch from online to persistent sessions without any additional setting or change in behavior. That is what I'm trying to point out here.

**Comment by ahus1** [comment]
> we would switch from online to persistent sessions without any additional setting or change in behavior.

@pedroigor - agreed, this is what I'll do. There are now two flag: One to enable storing the online sessions in the database, and another which disables the caches for online sessions as we want to performance test how the extra time to update cross-site caches relates to using the database only. 

This is still a draft as I wasn't able to complete all of my tests as the JBoss repository is currently down. I'll continue once it is back up again. Still, the last run looked quite good: https://github.com/keycloak/keycloak/actions/runs/8424835305

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    public PersistentSessionsChangelogBasedTransaction(KeycloakSession session, Cache<K, SessionEntityWrapper<V>> cache, RemoteCacheInvoker remoteCacheInvoker, SessionFunction<V> lifespanMsLoader, SessionFunction<V> maxIdleTimeMsLoader, boolean offline) {
+        super(session, cache, remoteCacheInvoker, lifespanMsLoader, maxIdleTimeMsLoader);
+        this.offline = offline;
+
+        if (!Profile.isFeatureEnabled(Profile.Feature.PERSISTENT_USER_SESSIONS)) {
+            throw new IllegalStateException("Persistent user sessions are not enabled");
+        }
+
+        if (Profile.isFeatureEnabled(Profile.Feature.PERSISTENT_USER_SESSIONS_NO_CACHE) && (cache.getName().equals(USER_SESSION_CACHE_NAME) || cache.getName().equals(CLIENT_SESSION_CACHE_NAME))) {
```
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java:48]: Why we differentiate between online and offline sessions here?

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java [hunk 2]
```diff
@@ -0,0 +1,1128 @@
...
+        }
+
+        @Override
+        public CrossDCMessageStatus getCrossDCMessageStatus(SessionEntityWrapper<UserSessionEntity> sessionWrapper) {
+            return CrossDCMessageStatus.SYNC;
+        }
+    }
+
+    public static UUID createClientSessionUUID(String userSessionId, String clientId) {
+        return UUID.nameUUIDFromBytes((userSessionId + clientId).getBytes(StandardCharsets.UTF_8));
```
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java:1126]: Is there some reason to generate ids like this? Can't we just use random?

**Review by mhajas — COMMENTED** [review]
Thanks @ahus1, the latest changes look good to me. I added two questions.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java (hunk 1 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java:48]: The "user session without a cache" will only be active for online sessions, not offline sessions. We could extend it in the next step to cover also offline sessions.

**Comment by ahus1** [comment]
The previous run completed green for all existing tests except for one with the old configuration, and with the new subset of tests. The one test was addressed in the last commit - it failed as it wasn't seeing the right log messages about preview features. https://github.com/keycloak/keycloak/actions/runs/8432316132/job/23100865619?pr=27977

In my personal repo, all old tests succeeded with the profiles enabled by default -  I added one more commit to a test-branch: https://github.com/ahus1/keycloak/actions/runs/8435548081

There I'm now switching this to ready-for-review.

**Comment by ahus1** [comment]
@pedroigor - it would be great if someone from the core-iam team could approve to merge this PR. As the main focus of the review, I suggest to ensure that it doesn't break the existing functionality, and that the current direction is the way we want to go. 

As a summary the main direction we're going here: Store online sessions in the existing user sessions table in the database, plus adding the option to not storing online sessions in the cache. 

There will be 2-3 more PRs before KC25 is released to optimize the newly added classes, for example the handling of database transactions, and removing some differences that now exist in the new code in handling offline and online sessions. We will also settle the question if online sessions should use a cache by running performance tests and gathering feedback from the community, so there will be only one feature in KC25 instead of two.

File: common/src/main/java/org/keycloak/common/Profile.java [hunk 3]
```diff
@@ -111,6 +111,9 @@ public enum Feature {
         HOSTNAME_V1("Hostname Options V1", Type.DEFAULT),
         //HOSTNAME_V2("Hostname Options V2", Type.DEFAULT, 2),
 
+        PERSISTENT_USER_SESSIONS("Persistent online user sessions across restarts and upgrades", Type.PREVIEW),
```
Comment (pedroigor) [inline common/src/main/java/org/keycloak/common/Profile.java:114]: Is it really preview at this stage? Doesn't make more sense as a experimental feature?

**Review by pedroigor — COMMENTED** [review]
@ahus1 The main point about my review is if it makes sense to have this feature as an `experimental` feature instead. 

I see some follow-ups in comments and in the code but not a blocker for me if the plan is to work on this incrementally. What makes me wonder if we can't deliver this in smaller chunks?

Also, what is the plan around authentication sessions as they are probably one of the most costly sessions as they are created whenever you reach the login page?

File: common/src/main/java/org/keycloak/common/Profile.java (hunk 3 above)
Comment (ahus1) [reply common/src/main/java/org/keycloak/common/Profile.java:114]: The main difference between experimental and preview is that preview is feature complete and that is provides some documentation. This is what we aim for in KC25. 

This PR already passes all the existing tests in the test suite, and serves as a drop-in-replacement for the existing user session handling. Given that, I think we're good to call it preview in the nightly build.  

Still, if this is the only remaining concern, I can change it experimental for now or in the next PR we're about to prepare. Let me know which way you decide.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java [hunk 4]
```diff
@@ -238,6 +238,9 @@ protected void addFeaturesOption(List<String> commands) {
             }
         }
 
+        commands.removeIf("--optimized"::equals);
+        commands.add("--http-relative-path=/auth");
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java:242]: Do we need these two options in the testsuite?

File: testsuite/integration-arquillian/tests/base/testsuites/persistent-sessions-suite
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/testsuites/persistent-sessions-suite:1]: I would say we should include more tests from [oauth](https://github.com/keycloak/keycloak/tree/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth) package as most of them are using sessions for something.

I remember my changes caused failures in: 
`RefreshTokenTest`
`OfflineTokenTest`

**Review by mhajas — DISMISSED** [review]
Thank you @ahus1! I checked the code for breaking changes for the old behavior and haven't found anything. 

I added two additional comments in the tests but if it results in some changes we can include it in one of follow-up PRs.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java (hunk 4 above)
Comment (ahus1) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java:242]: Once we change features, we need to rebuild the image. Once you rebuild the image, you need to-add the build time parameters, which was the `http-relative-path`. There is a similar snippet of code in the same file.  

Apparently this is the first time we're changing features on the command line?

So TL;DR: Yes.

File: testsuite/integration-arquillian/tests/base/testsuites/persistent-sessions-suite
Comment (ahus1) [reply testsuite/integration-arquillian/tests/base/testsuites/persistent-sessions-suite:1]: Ok. added the following tests in the next commit:

```
RefreshTokenTest
OfflineTokenTest
AccessTokenTest
LogoutTest
```

**Comment by ahus1** [comment]
@mhajas - I've pushed additional commits to address @pedroigor's comment (the new feature flags are now "experimental"), the new tests are added, and I refactored the code to make it more descriptive why the optimized flag is removed and another is added.

Please re-review and approve when you have the time. Given that @pedroigor's comment has been addressed, it should then be read to be merged.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@stianst @ahus1 Up to you to decide whether we should have this one merged or in a separate branch.

**Review by mhajas — APPROVED** [review]
Thank you @ahus1

**Pull request merged**